doc/rxvtd.1.pod
=head1 NAME

@@RXVT_NAME@@d - @@RXVT_NAME@@ terminal daemon

=head1 SYNOPSIS

B<@@RXVT_NAME@@d> [-q|--quiet] [-o|--opendisplay] [-f|--fork] [-m|--mlock] [-e|--eval I<perlstring>]

B<@@RXVT_NAME@@d> -q -o -f    # for .xsession use

=head1 DESCRIPTION

This manpage describes the @@RXVT_NAME@@d daemon, which is the same vt102
terminal emulator as @@RXVT_NAME@@, but runs as a daemon that can open
multiple terminal windows within the same process.

You can run it from your X startup scripts, for example, although it is
not dependent on a working DISPLAY and, in fact, can open windows on
multiple X displays on the same time.

Advantages of running a @@RXVT_NAME@@ daemon include faster creation time
for terminal windows and a lot of saved memory.

The disadvantage is a possible impact on stability - if the
main program crashes, all processes in the terminal windows are
terminated. For example, as there is no way to cleanly react to abnormal
connection closes, C<xkill> and server resets/restarts will kill the
B<@@RXVT_NAME@@d> instance including all windows it has opened.

=head1 OPTIONS

B<@@RXVT_NAME@@d> currently understands a few options only. Bundling of
options is not yet supported.

=over

=item B<-q>, B<--quiet>

Normally, B<@@RXVT_NAME@@d> outputs the message C<< rxvt-unicode daemon
listening on <path> >> after binding to its control socket. This option
will suppress this message (errors and warnings will still be logged).

=item B<-o>, B<--opendisplay>

This forces B<@@RXVT_NAME@@d> to open a connection to the current
C<$DISPLAY> and keep it open.

This is useful if you want to bind an instance of B<@@RXVT_NAME@@d> to
the lifetime of a specific display/server. If the server does a reset,
B<@@RXVT_NAME@@d> will be killed automatically.

=item B<-f>, B<--fork>

This makes B<@@RXVT_NAME@@d> fork after it has bound itself to its control
socket.

=item B<-m>, B<--mlock>

This makes B<@@RXVT_NAME@@d> call mlockall(2) on itself. This locks
B<@@RXVT_NAME@@d> in RAM and prevents it from being swapped out to disk,
at the cost of consuming a lot more memory on most operating systems.

Note: In order to use this feature, your system administrator must have set
your user's RLIMIT_MEMLOCK to a size greater than or equal to the size of the
B<@@RXVT_NAME@@d> binary (or to unlimited). See F</etc/security/limits.conf>.

Note 2: There is a known bug in glibc (possibly fixed in 2.8 and later
versions) where calloc returns non-zeroed memory when mlockall is in
effect. If you experience crashes or other odd behaviour while using
--mlock, try it without it.

=item B<-e>, B<--eval> I<perlstring>

Evaluate the given perl code after basic initialisation (requires perl
support to be enabled when compiling B<@@RXVT_NAME@@d>).

This can be used for example to configure the internal perl interpreter,
which is shared between all terminal instances, or create additional
listening sockets for additional protocols.

The code is currently executed I<before> creating the normal listening
sockets: this might change in future versions.

=back

=head1 EXAMPLES

This is a useful invocation of B<@@RXVT_NAME@@d> in a F<.xsession>-style
script:

   @@RXVT_NAME@@d -q -f -o

This waits till the control socket is available, opens the current display
and forks into the background. When you log-out, the server is reset and
B<@@RXVT_NAME@@d> is killed.

=head1 ENVIRONMENT

=over

=item B<RXVT_SOCKET>

Both B<@@RXVT_NAME@@c> and B<@@RXVT_NAME@@d> use the environment
variable F<RXVT_SOCKET> to create a listening socket and to contact
the @@RXVT_NAME@@d, respectively. If the variable is missing then
F<<< $HOME/.urxvt/urxvtd-I<< <nodename> >> >>> is used.

=item B<LISTEN_PID>, B<LISTEN_FDS>, B<LISTEN_FDNAMES>

if these environment variables are set, @@RXVT_NAME@@d checks for a file
descriptor passed by the service manager via systemd native interface. If
one is provided, @@RXVT_NAME@@d uses it as its listening socket instead of
creating one as specified above.

=item B<DISPLAY>

Only used when the C<--opendisplay> option is specified. Must contain a
valid X display name.

=back

=head1 SEE ALSO

@@RXVT_NAME@@(7), @@RXVT_NAME@@c(1)


Changes
rxvt-unicode changelog <= google-friendly title

TODO: xcopyarea pass broken, fix and improve
TODO: event mechanism that replaces on_user_command with something more scalable.
TODO: slipwheeling needs a keyboard grab to get key release events.
TODO: overlays collide with the way the out-of-focus cursor is being drawn
TODO: "slow" rendering mode for bidi and scripts, pango!
TODO: read property sequence is broken with respect to utf-8 etc.
TODO: slow drag selection on remote display - other things seem fast (rxvt illness, motion hints?)
TODO: better resource handling for tabbar in -pe tabbed.
WISH: investigate -pe tabbed -g 80x25 being 23 not 24 or 25 lines
      => when setting the constraints, the wm might (correctly)
         resize the outer window, which might result in a smaller window overall.
         or something like that. guys, send me a patch.
WISH: coloured pastebin
WISH: load system-wide config file even if we don't have one
WISH: look into XAddConnectionWatch, does anybody need that?
DUMB: support tex fonts
TODO: perl-shell-window? perl-unix-socket?
TODO: command line editing when icanon?
TODO: split ROW into a ROW_fast (0..total*2-1) and ROW macros?
TODO: catch exceptions when calling perl hooks
TODO: xdbe
TODO: optimise bg reeval for resize for common patterns (pixmap bigger than window?)
TODO: pack rxvt_fatal messages into the exception object
TODO: load must not cache, but global image cache must be cleaned
TODO: description into --help output?
TODO: rxvt -help => (see rxvt-XXX(1))?
TODO: xor_rect should shift right
TODO IMPL: recalc bg only when not fully covering (no alpha, repeat|extend etc..)

TODO IMPL: colour-change event with index
TODO IMPL: recalc bg always on bg colour change

TODO: document typical actions in rxvt.1.pod keysyms
TODO: warn with a graphical message when env has been modified
TODO: c&c perl socket via daemon-ext mechanism
TODO: simplify extension metainfo cache, cache on disk
TODO: URxvt::Ext::Name installs urxvt ext name and provides pod/manpage for URxvt::Ext::Name
TODO: üpixel droppins idenrasm,ll,scrollup
TODO: cuu/cud and probably others default to 1 when arg is 0, not just missing, in xterm/vt102, but not in rxvt
TODO: clean up the resource vs. long option mess, especially in the manpage.

TODO: osc 20 does not autoload background ext
TODO: think about daemon delaying urxvt client exit.
IDEA: use harfbuzz to render combining sequences, maybe also complex scripts

TODO: https://github.com/bbidulock/icewm/issues/492 urxvt wrongfully(?) needa configurenotify after mapwindow? also, black border?
	- enable confirm-paste extension by default, can be disabled by e.g.
          URxvt.perl-ext: -confirm-paste
        - Key bindings specified with the -keysym option or with resources of
          the form URxvt*keysym work again. Such resources, as opposed to
          resources of the form URxvt.keysym, were not officially allowed but
          accepted, so ignoring them is considered a regression.
        - new option/resource -rm/rewrapMode, which sets the line rewrap
          mode to auto, never or always.
        - new option/resource -dpb/disablePasteBrackets, which can be used to
          disable bracketed paste mode sequences, as more and more programs
          abuse these sequences (typically in the name of security, but these
          sequences are not improving security). Also, bracketed paste mode
          is now under frills.
	- if urxvt cannot detect a good estimate for font width (because a font does
          not contain latin characters at all for example), it will now fall
          back to xft's max_advance_width, which works well with some fonts. It
          currently will warn about this.
	- do not render variation selectors in the default font as boxes, treat
          them as zero-width characters.
        - change combining character architecture to allow fonts to handle
          combining sequences on their own. This could allow future renderers
          to be more intelligent about these sequences.
        - use the previous enhancement to allow unicode characters outside of the
          BMP to be rendered even without --enable-unicode3, as long as an
          xft with the character is configured.
          It's recommended to stay with disabled unicode3 even more now.
        - freetype/xft have been verified to return nonsensical offsets for
          combining characters. urxvt tries to compensate to some extent,
          but diacritics will be slightly off in many cases, and far off in
          some cases (e.g. input mono).
        - when checking for overlap/careful handling, take non-basefont
          character shifts into account.
        - extensions can now request autoloading on OSC and Perl OSC sequences,
          implement this for the background, clipboard-osc and overlay-osc
          extensions.
        - the background extension will now be autoloaded when the OSC 20 or 705
          sequence is received.
	- The view_start perl method now scrolls to the bottom also when the
	  argument is greater than 0.
	- eval scroll actions now scroll to the bottom when scrolling down and
	  there are less lines to scroll than the ones requested (reported by
	  Stephen Talley).
	- allow perl-ext and perl-ext-common to override autoloaded extensions.
          this allows tabbed to disable itself in subtabs, avoiding recursive
          loading.
	- update precompose table to unicode 13.0.
        - update charset conversion tables to perl 5.32 (this only affected
          cs_koi8_u_from_unicode).
        - unicode characters > 0x100000 were not handled correctly w.r.t.
          font rendering.
        - replace bitstream vera by dejavu sans, as it has replaced the former
          almost everywhere.
        - shave off over 1K of code size from urclock by replacing the sine table
          with a slightly less accurate fixed point approximation.

src/rxvtd.C
/*----------------------------------------------------------------------*
 * File:	rxvtd.C
 *----------------------------------------------------------------------*
 *
 * All portions of code are copyright by their respective author/s.
 * Copyright (c) 2003-2007 Marc Lehmann <schmorp@schmorp.de>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *----------------------------------------------------------------------*/

#include "../config.h"
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>

#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/un.h>

#if defined(ENABLE_FRILLS) && defined(_POSIX_MEMLOCK) && _POSIX_MEMLOCK > 0
# define ENABLE_MLOCK 1
#endif

#if ENABLE_MLOCK
# include <sys/mman.h>
#endif

#include <errno.h>

#include <libptytty.h>

#include "rxvt.h"
#include "rxvtdaemon.h"

struct server : rxvt_connection
{
  log_callback log_cb;
  getfd_callback getfd_cb;

  void read_cb (ev::io &w, int revents); ev::io read_ev;
  void log_msg (const char *msg);
  int getfd (int remote_fd);

  server (int fd)
  {
    read_ev.set <server, &server::read_cb> (this);
    log_cb.set  <server, &server::log_msg> (this);
    getfd_cb.set<server, &server::getfd>   (this);

    this->fd = fd;
    fcntl (fd, F_SETFD, FD_CLOEXEC);
    fcntl (fd, F_SETFL, 0);
    read_ev.start (fd, ev::READ);
  }

  void err (const char *format = 0, ...);
};

struct unix_listener
{
  int fd;

  void accept_cb (ev::io &w, int revents); ev::io accept_ev;

  unix_listener (int fd);
  static int open (const char *sockname);
};

unix_listener::unix_listener (int fd) : fd (fd)
{
  accept_ev.set<unix_listener, &unix_listener::accept_cb> (this);
  fcntl (fd, F_SETFD, FD_CLOEXEC);
  fcntl (fd, F_SETFL, O_NONBLOCK);
  accept_ev.start (fd, ev::READ);
}

int unix_listener::open (const char *sockname)
{
  sockaddr_un sa;

  if (strlen (sockname) >= sizeof(sa.sun_path))
    {
      fputs ("socket name too long, aborting.\n", stderr);
      exit (EXIT_FAILURE);
    }

  int fd;
  if ((fd = socket (AF_UNIX, SOCK_STREAM, 0)) < 0)
    {
      perror ("unable to create listening socket");
      exit (EXIT_FAILURE);
    }

  sa.sun_family = AF_UNIX;
  strcpy (sa.sun_path, sockname);

  unlink (sockname);

  mode_t omask = umask (0077);

  if (bind (fd, (sockaddr *)&sa, sizeof (sa)))
    {
      perror ("unable to bind listening socket");
      exit (EXIT_FAILURE);
    }

  umask (omask);

  if (listen (fd, 5))
    {
      perror ("unable to listen on socket");
      exit (EXIT_FAILURE);
    }

  return fd;
}

void unix_listener::accept_cb (ev::io &w, int revents)
{
  int fd2 = accept (fd, 0, 0);

  if (fd2 >= 0)
    new server (fd2);
}

int server::getfd (int remote_fd)
{
  send ("GETFD");
  send (remote_fd);
  return ptytty::recv_fd (fd);
}

void server::log_msg (const char *msg)
{
  send ("MSG"), send (msg);
}

void server::err (const char *format, ...)
{
  if (format)
    {
      char err[1024];

      va_list ap;
      va_start (ap, format);
      vsnprintf (err, 1024, format, ap);
      va_end (ap);

      log_msg (err);
    }

  send ("END"), send (0);
  close (fd);
  delete this;
}

void server::read_cb (ev::io &w, int revents)
{
  auto_str tok;

  if (recv (tok))
    {
      if (!strcmp (tok, "NEW"))
        {
          stringvec *argv = new stringvec;
          stringvec *envv = new stringvec;

          for (;;)
            {
              if (!recv (tok))
                return err ();

              if (!strcmp (tok, "END"))
                break;
              else if (!strcmp (tok, "ENV") && recv (tok))
                envv->push_back (strdup (tok));
              else if (!strcmp (tok, "ARG") && recv (tok))
                argv->push_back (strdup (tok));
              else
                return err ("protocol error: unexpected NEW token.\n");
            }

          {
            rxvt_term *term = new rxvt_term;

            term->log_hook = &log_cb;
            term->getfd_hook = &getfd_cb;

            bool success = true;

            try
              {
                term->init (argv, envv);
              }
            catch (const std::exception &e)
              {
                log_msg (e.what());
                success = false;
              }

            term->log_hook = 0;

            chdir ("/"); // init might change to different working directory

            if (!success)
              term->destroy ();

            send ("END"); send (success ? 1 : 0);
          }
        }
      else if (!strcmp (tok, "QUIT"))
        _exit (0);
      else
        return err ("protocol error: request '%s' unsupported.\n", (char *)tok);
    }
  else
    return err ();
}

#ifdef ENABLE_FRILLS
enum {
  SD_LISTEN_FDS_START = 3,
};

// https://www.freedesktop.org/software/systemd/man/sd_listen_fds.html
// https://github.com/systemd/systemd/blob/main/src/libsystemd/sd-daemon/sd-daemon.c
static int get_listen_fds ()
{
  const char *listen_pid = getenv ("LISTEN_PID");
  if (!listen_pid)
    return 0;

  char *end;
  errno = 0;
  long pid = strtol (listen_pid, &end, 10);
  if (errno || end == listen_pid || *end)
    return -1;

  if (getpid () != pid)
    return 0;

  const char *listen_fds = getenv ("LISTEN_FDS");
  if (!listen_fds)
    return 0;

  errno = 0;
  long n = strtol (listen_fds, &end, 10);
  if (errno || end == listen_fds || *end)
    return -1;

  if (n <= 0 || n > INT_MAX - SD_LISTEN_FDS_START)
    return -1;

  unsetenv ("LISTEN_PID");
  unsetenv ("LISTEN_FDS");
  unsetenv ("LISTEN_FDNAMES");

  return n;
}
#endif

int
main (int argc, char *argv[])
{
  try
    {
      ptytty::init ();
    }
  catch (const std::exception &e)
    {
      fputs (e.what (), stderr);
      return EXIT_FAILURE;
    }

  static char opt_fork, opt_opendisplay, opt_quiet;
#if ENABLE_PERL
  static char *opt_eval;
#endif
#if ENABLE_MLOCK
  static char opt_lock;
#endif

  for (int i = 1; i < argc; i++)
    {
      if (!strcmp (argv [i], "-f") || !strcmp (argv [i], "--fork"))
        opt_fork = 1;
      else if (!strcmp (argv [i], "-o") || !strcmp (argv [i], "--opendisplay"))
        opt_opendisplay = 1;
      else if (!strcmp (argv [i], "-q") || !strcmp (argv [i], "--quiet"))
        opt_quiet = 1;
#if ENABLE_MLOCK
      else if (!strcmp (argv [i], "-m") || !strcmp (argv [i], "--mlock"))
        opt_lock = 1;
#endif
#if ENABLE_PERL
      else if (!strcmp (argv [i], "-e") || !strcmp (argv [i], "--eval"))
        opt_eval = argv [++i];
#endif
      else
        {
          rxvt_log ("%s: unknown option '%s', aborting.\n", argv [0], argv [i]);
          return EXIT_FAILURE;
        }
    }

  rxvt_init ();

#if ENABLE_PERL
  if (opt_eval)
    {
      rxvt_perl.init ();
      rxvt_perl.eval (opt_eval);
    }
#endif

  // optionally open display and never release it.
  if (opt_opendisplay)
    if (const char *dpy = getenv ("DISPLAY"))
      displays.get (dpy ? dpy : ":0"); // move string logic into rxvt_display maybe?

  int fd;
#ifdef ENABLE_FRILLS
  int n = get_listen_fds ();
  if (n > 1)
    {
      fputs ("received multiple file descriptors, aborting.\n", stderr);
      exit (EXIT_FAILURE);
    }
  else if (n == 1)
    {
      fd = SD_LISTEN_FDS_START;
      if (!opt_quiet)
        {
          printf ("rxvt-unicode daemon listening on fd.\n");
          fflush (stdout);
        }
    }
  else
#endif
    {
      char *sockname = rxvt_connection::unix_sockname ();
      fd = unix_listener::open (sockname);
      if (!opt_quiet)
        {
          printf ("rxvt-unicode daemon listening on %s.\n", sockname);
          fflush (stdout);
        }
      free (sockname);
    }

  unix_listener l (fd);

  chdir ("/");

  pid_t pid = 0;
  if (opt_fork)
    {
      pid = fork ();
    }

#if ENABLE_MLOCK
  // Optionally perform an mlockall so this process does not get swapped out.
  if (opt_lock && !pid)
    if (mlockall (MCL_CURRENT | MCL_FUTURE) < 0)
      perror ("unable to lock into ram");
#endif

  if (opt_fork)
    {
      if (pid < 0)
        {
          rxvt_log ("unable to fork daemon, aborting.\n");
          return EXIT_FAILURE;
        }
      else if (pid > 0)
        _exit (EXIT_SUCCESS);

      ev_loop_fork (EV_DEFAULT_UC);
    }

  ev_run ();

  return EXIT_SUCCESS;
}

